libsolidity/analysis/TypeChecker.cpp

```diff
@@ -697,11 +697,16 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 	}
 	for (ASTPointer<VariableDeclaration> const& var: _function.parameters() + _function.returnParameters())
 	{
-		if (
-			type(*var)->category() == Type::Category::Mapping &&
-			!type(*var)->dataStoredIn(DataLocation::Storage)
-		)
-			m_errorReporter.typeError(var->location(), "Mapping types can only have a data location of \"storage\".");
+		if (type(*var)->category() == Type::Category::Mapping)
+		{
+			if (!type(*var)->dataStoredIn(DataLocation::Storage))
+				m_errorReporter.typeError(var->location(), "Mapping types can only have a data location of \"storage\"." );
+			else if (
+				!isLibraryFunction &&
+				_function.visibility() > FunctionDefinition::Visibility::Internal
+			)
+				m_errorReporter.typeError(var->location(), "Type is required to live outside storage.");
```

Comment:
I think this message can be improved, something like `mapping type can only be used as argument in internal or library functions`